yt/utilities/answer_testing/framework.py

```diff
@@ -63,6 +81,332 @@
+        dd = ds.all_data()
+        dens_enzo = dd["Density"].copy()
+        for f in sorted(ds.field_list):
+            ff = f[1]
+            if not ff.endswith("_Density"):
+                continue
+            start_strings = ["Electron_", "SFR_", "Forming_Stellar_",
+                             "Dark_Matter", "Star_Particle_"]
+            if any([ff.startswith(ss) for ss in start_strings]):
+                continue
+            dens_enzo -= dd[f]
+        delta_enzo = np.abs(dens_enzo / dd["Density"])
+        np.testing.assert_almost_equal(delta_yt, delta_enzo)
+
+    def simulated_halo_mass_function_test(self, ds, finder):
+        hc = HaloCatalog(data_ds=ds, finder_method=finder)
+        hc.create()
+        hmf = HaloMassFcn(halos_ds=hc.halos_ds)
+        result = np.empty((2, hmf.masses_sim.size))
+        result[0] = hmf.masses_sim.d
+        result[1] = hmf.n_cumulative_sim.d
+        return result
+
+    def analytic_halo_mass_function_test(self, ds, fit):
+        hmf = HaloMassFcn(simulation_ds=ds, fitting_function=fit)
+        result = np.empty((2, hmf.masses_analytic.size))
+        result[0] = hmf.masses_analytic.d
+        result[1] = hmf.n_cumulative_analytic.d
+        return result
+
+    def small_patch_amr(self, ds, field, weight, axis, ds_obj):
+        hex_digests = {} 
+        # Grid hierarchy test
+        gh_hd = self.grid_hierarchy_test(ds)
+        hex_digests['grid_hierarchy'] = gh_hd
+        # Parentage relationships test
+        pr_hd = self.parentage_relationships_test(ds)
+        hex_digests['parentage_relationships'] = pr_hd
+        # Grid values, projection values, and field values tests
+        gv_hd = self.grid_values_test(ds, field)
+        hex_digests['grid_values'] = gv_hd
+        fv_hd = self.field_values_test(ds, field, ds_obj)
+        hex_digests['field_values'] = fv_hd
+        pv_hd = self.projection_values_test(ds, axis, field, weight, ds_obj)
+        hex_digests['projection_values'] = pv_hd
+        return hex_digests
+
+    def big_patch_amr(self, ds, field, weight, axis, ds_obj):
+        hex_digests = {} 
+        # Grid hierarchy test
+        gh_hd = self.grid_hierarchy_test(ds)
+        hex_digests['grid_hierarchy'] = gh_hd
+        # Parentage relationships test
+        pr_hd = self.parentage_relationships_test(ds)
+        hex_digests['parentage_relationships'] = pr_hd
+        # Grid values, projection values, and field values tests
+        gv_hd = self.grid_values_test(ds, field)
+        hex_digests['grid_values'] = gv_hd
+        ppv_hd = self.pixelized_projection_values_test(ds, axis, field, weight, ds_obj)
+        hex_digests['pixelized_projection_values'] = ppv_hd 
+        return hex_digests
+
+    def generic_array_test(self, func, args=None, kwargs=None):
+        if args is None:
+            args = []
+        if kwargs is None:
+            kwargs = {}
+        return func(*args, **kwargs)
```

Comment:
Are you sure this is doing the same thing as the old `generic_array_test`? It doesn't appear to be saving or comparing answers, for example.

Reply:
Oh I see, I guess it just needs a usage example to be clearer, but I understand how this works now. You pass it a function and arguments and keyword arguments to that function and it saves the result of that function.